File: oss-2022/scala/src/main/scala/BikeTypeJob.scala
package io.pipekit.oss.spark.argo

import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession

object BikeTypeJob {
  val fileLocation = "/app/df_1_year.csv"

  def main(args: Array[String]): Unit = {
    Logger.getLogger("org").setLevel(Level.ERROR)

    val spark = SparkSession
      .builder
      .appName("BikeTypeJob")
      .getOrCreate()

    val rdd = spark.sparkContext.textFile(fileLocation).map(Bike.toBikeRide)

    val bikeTypeStats = rdd
      .map(x => (x.bikeType, 1))
      .reduceByKey((x, y) => x + y)
      .sortBy(x => x._2, ascending = false).collect()

    for (elem <- bikeTypeStats) {
      println(s"BikeType ${elem._1}, count ${elem._2}")
    }

    spark.stop()
  }

}

File: oss-2022/python/bikeTypeJob.py
import os
from pyspark.sql import SparkSession


FILE_LOCATION = os.environ.get("FILE_LOCATION")
AWS_ACCESS_KEY_ID = os.environ.get("AWS_ACCESS_KEY_ID")
AWS_SECRET_KEY = os.environ.get("AWS_SECRET_KEY")

spark = SparkSession.builder.appName("bike-type-job").getOrCreate()
spark._jsc.hadoopConfiguration().set("fs.s3a.access.key", AWS_ACCESS_KEY_ID)
spark._jsc.hadoopConfiguration().set("fs.s3a.secret.key", AWS_SECRET_KEY)

rdd = spark.read.csv(FILE_LOCATION, header=True).rdd

result = rdd \
    .map(lambda fields: (fields["rideable_type"], 1)) \
    .reduceByKey(lambda x, y: x + y) \
    .sortBy(lambda x: x[1], ascending=False) \
    .collect()

for element in result:
    print("BikeType %s, count %d" %(element[0], element[1]))


spark.stop()
